### DISCUSSION TITLE
Switch from flake8/black to ruff

### DISCUSSION THREAD

**Discussion by uriyyo** [sqlalchemy/sqlalchemy#10683] [Ideas]
Ruff is an extremely fast linter/formatter, much faster than flake8 and black. 

![image](https://github.com/sqlalchemy/sqlalchemy/assets/32038156/ec0cff93-596d-4e35-8e9a-86c3f1d50e08)

Many big open-source projects switch to ruff (like fastapi and pandas).

I can submit PR if you interested in migration.

**Comment by CaselIT** [comment]
Hi,

Last time this was looked there were some extensions or options missing from ruff.

Also honestly the time taken is not really a pressing issue. The pipeline that runs the all the linting takes a couple of minutes, including installs, on the not really fast GitHub machines https://github.com/sqlalchemy/sqlalchemy/actions/runs/6963296347/job/18948630487

**Reply by CaselIT** [reply]
@zzzeek not sure if you already did some experiments with ruff

**Reply by zzzeek** [reply]
I did, it's a complete non-starter for us as we use many flake8 extensions including some that I wrote myself.  

flake8 works fine for us and it's plenty fast.

**Comment by devkral** [comment]
I have some other arguments for ruff as linter:

They have absorbed a lot of linting projects (new development happens in ruff) and most (if not all) flake8 extensions.
This means you have much more choice without installing new linting libraries.

Given the popularity of ruff many of them will be abandoned and you get python version problems not to mention the security aspect if some small linting extension becomes rogue because the maintainer doesn't care anymore.

Upgrading from flake8 is also really easy because it uses nearly the same syntax (they got inspired by it). So yeah, I would really look forward to see ruff as linter (and can provide patches) but I don't want any linter wars ;).

**Reply by CaselIT** [reply]
it seems that we have one. I don't think flake will be replaced by ruff. if python can support 5+ package manager (pip, poetry, rye, uv, conda, pdm, etc) I don't see why it can't have two linter.

I don't understand why people care what linter a library uses

**Reply by devkral** [reply]
It is quite simple:

if I contribute I get no automatic formatting while typing and have to configure my IDE to not apply the wrong formatting (from ruff it is easily picked up (if I would configure flake8 the result would depend on my system libraries)). Of course this is just a nuisance but easier for some (including me) would be ruff.

**Reply by devkral** [reply]
And second don't forget precommit. I have no good internet connection and having many deps is slow for me.

**Reply by CaselIT** [reply]
would a tox env help with formatting? There's already one to check the linting (`lint`)

**Comment by zzzeek** [comment]
Ruff has no support for plugins and we rely heavily on several plugins, one of which is part of the SQLAlchemy project that we maintain.   Ruff has no advantages of any kind that I'm familiar with except for better performance, and flake8 is perfectly fine even for a large codebase like that of SQLAlchemy.  So that's end of story for now.

**Comment by zzzeek** [comment]
SQLAlchemy has no need or plans to switch out of flake8 barring some major ecosystem changes and introducing a rust-based tool would only introduce lots of complexities, if it even ever supported all of our plugins, so closing this.

**Reply by devkral** [reply]
I used flake8 before ruff ;). Amazing tool anyway.

**Reply by devkral** [reply]
Would you mind a PR adding settings in pyproject.toml preventing ruff formatting and linting, until sqlalchemy wants it?
As a reminder, so people don't apply it here.

**Reply by CaselIT** [reply]
I think it would be ok also if the ruff config were told to format the same as black (not sure if it has a config option "look-at-the-black-config=true" or similar)
